procman.1
.\" Manpage for procman.

.TH MAN 1 "16 June 2020" "1.0" "procman man page"

.SH NAME
procman \- list active processes

.SH SYNOPSIS
\fBprocman show\fR [\fB-s\fR|\fB--sort\fI category\fR] [\fB-E\fR|\fB--regex\fR] [\fB-x\fR|\fB--exp\fR\fI expression\fR] [\fB-f\fR|\fB--filter\fR\fI filter\fR]
.br
\fBprocman kill\fR \fIname\fR|\fIpID\fR
.br
\fBprocman details\fR \fIname\fR|\fIpID\fR
.br
\fBprocman\fR [\fBhelp\fR]

.SH DESCRIPTION
procman is a shell program that allows a user to easily monitor and manipulate active processes on their system.

.SH OPTIONS
Running \fBprocman\fR or \fBprocman help\fR will show the usage of this function, similar to the synopsis above. Running \fBprocman show\fR without any additional arguments will show a list of all active processes in ascending order.
.SS Options for show
.TP 7
\fB-s\fI category\fR, \fB--sort\fI category\fR
sort the list of active processes in descending order by \fIcategory\fR. Category must be either \fBpid\fR, \fBmem\fR, or \fBcpu\fR. Cannot be used with \fB-x\fR or \fB--exp\fR, however can be used with \fB-f\fR or \fB--filter\fR.
.br

.TP
\fB-f\fI filter\fR, \fB--filter\fI filter\fR
filter the list of active processes by \fIfilter\fR. Finds only entries containing \fIfilter\fR by default. Cannot be used with \fB-x\fR or \fB--exp\fR, however can be used with \fB-s\fR or \fB--sort\fR.

.TP
\fB-E -f\fR\fI filter\fR, \fB--regex --filter\fI filter\fR
Makes filter look instead for entries matching \fIfilter\fR as if it were a regular expression. Cannot be used with \fB-x\fR or \fB--exp\fR, however can be used with \fB-s\fR or \fB--sort\fR. Must be used with \fB-f\fR or \fB--filter\fR.

.TP
\fB-x\fI expression\fR, \fB--exp\fI expression\fR
Looks for entries that satisfy the expression represented by \fIexpression\fR, which must be in the form \fIcolumn\fR>\fIvalue\fR or \fIcolumn\fR<\fIvalue\fR where \fIcolumn\fR is either \fBpid\fR, \fBcpu\fR, or \fBmem\fR and \fIvalue\fR is either an integer or floating-point decimal. Ensure that \fIexpression\fR is wrapped in double quotes. \fB<=\fR and \fB>=\fR operators are also permitted. Cannot be used with \fB-E\fR, \fB-s\fR, or \fB-f\fR flags.

Ex: \fBprocman show -x "cpu>0.0"\fR will print all processes with cpu usage greater than 0.0%.

.SS Options for kill
.TP
\fBname\fR
Kills the active process with name \fIname\fR.

.TP
\fBpID\fR
Kills the active process with ID \fIpID\fR.

.SS Options for details
.TP
\fBname\fR
shows the status of the process with name \fIname\fR and the command that is running it.

.TP
\fBpID\fR
shows the status of the process with ID \fIpID\fR and the command that is running it.

.SH BUGS
No known bugs.
.SH AUTHORS
Jason Klein
.br
Darius Kianersi
.br
Kaushik Siruvuri


procman.sh
#!/bin/bash

procmanShow() {
    {
        INPUT=$@
        COMMAND="ps aux" # command to show output of

	    # convert user input to ps arg
	    declare -A sorts=( ["pid"]="pid" ["cpu"]="pcpu" ["mem"]="rss")

        # convert user input to column number
        declare -A cols=( ["pid"]=2 ["cpu"]=3 ["mem"]=4)

        # sorting #
        if [[ $INPUT =~ -s|--sort ]]; then
            SORT="$(grep -oP "(?:(?<=-s )|(?<=--sort ))\w+" <<< $INPUT)"
	    SORT="${sorts[$SORT]}"
	  
            if [[ "$SORT" != "" ]]; then
                COMMAND="ps aux --sort -$SORT"
	    else
		COMMAND="echo Invalid category for sort flag. Use either pid, mem, or cpu."
            fi
        fi

        # check for regex flag #
        REGEX=false
        if [[ $INPUT =~ -E|--regex ]]; then
            # remove flag from INPUT #
            INPUT="$(sed -e 's/-E //g' -e 's/--regex //g' <<< $INPUT)"
            REGEX=true
        fi
	
        # check for expression flag #
        EXPRESSION=false
        if [[ $INPUT =~ -x|--exp ]]; then 
            EXPRESSION=true
        fi

	# check for filter flag #
	FILFLAG=false
	if [[ $INPUT =~ -f|--filter ]]; then
	    FILFLAG=true
	fi

        # filtering #
        if $FILFLAG ; then
            FILTER="$(grep -oP '((?<=-f )|(?<=--filter ))(\w+(..?[0-9\.]+)?)' <<< $INPUT)"

            if $REGEX ; then
                COMMAND="$COMMAND | grep -E $FILTER"

            elif [[ "$FILTER" != "" ]]; then
                COMMAND="$COMMAND | grep $FILTER"

            # handle no filter #
            else
                COMMAND="echo No filter specified. Refer to ./procman.sh help for more guidance."
            fi
        fi
 
	if [[ $INPUT =~ -x|--exp ]]; then
	    if [[ "$SORT" != "" ]] || $FILFLAG ; then
                COMMAND="echo Cannot use sort or filter flag with expression flag. Refer to ./procman.sh help for more guidance."

            else
		EXP="$(grep -oP '((?<=-x )|(?<=--exp ))(\w+(..?[0-9\.]+)?)' <<< $INPUT)"
                COL="$(grep -o -E "^[a-z]+" <<< $EXP)" # column name
                TYPE="${sorts[$COL]}" # ps arg
                VALUE="$(grep -oP '[0-9\.]+$' <<< $EXP)" # value to compare
                OPERATOR="$(grep -oP '(<=|<|>=|>)' <<< $EXP)" # operator to use
                COLNUM="${cols[$COL]}" # column number
		    
		# throw error for invalid arg(s)
		if [ -z $COL ] || [ -z $TYPE ] || [ -z $VALUE ] || [ -z $OPERATOR ] || [ "$COL$OPERATOR$VALUE" != $EXP ]; then
		    COMMAND="echo 'Invalid argument(s) provided. Refer to ./procman.sh help for more guidance.'"
		else
                    let count=0	
                    if [[ $OPERATOR =~ ">" ]]; then
                        TYPE="-$TYPE"
                    fi
                    # count number of instances
                    while read line; do
			    
                        if (( $(echo $line $OPERATOR $VALUE |bc -l) )); then # compare with argument
                            count=$((count+1))
                        fi
                    done < <(ps aux --sort=$TYPE | sed '1d;$d' | awk "{print \$$COLNUM}")
                    count=$((count+1)) # increment to account for table heading
                    COMMAND="ps aux --sort=$TYPE | head -n $count"
	        fi
            fi
	fi

    # hide all output #
    } &> /dev/null

    # add table header if removed #
    if [[ "$COMMAND" =~ grep ]]; then
        ps aux | sed "1q"
    fi

    # eval $COMMAND
    eval $COMMAND | 
        while read LINE; do
            echo "$LINE"
        done
}

procmanKill() {
    # check if arg is a number #
    if [[ $1 =~ ^-?[0-9]+$ ]]; then
	    kill $1
    else
	    pkill $1
    fi
}

procmanDetails() {
    # check if arg is a number #
    if [[ $1 =~ ^-?[0-9]+$ ]]; then
	    pid=$1
    else
	    pid=$(pidof $1 | awk '{print $1}')
    fi
    cat /proc/$pid/status
    echo
    cat /proc/$pid/cmdline
    echo
}

procmanHelp() {
    man ./procman.1
}

case $1 in 
    "show")
        procmanShow "$@"
        # maybe replace with procmanShow $@
        # $2 is --sort or -s, $3 is property, $4 is --filter or -f, $5...$# is what to filter by
        ;;

    "kill")
        procmanKill "$2"
        ;;

    "details")
	procmanDetails "$2"
	;;

    "help")
        procmanHelp
        ;;

    *)
        # default case, help script here
	echo "Command not found. Enter './procman.sh help' for a list of commands"
        ;;
esac
